python/ray/_private/runtime_env/packaging.py

```diff
@@ -409,19 +447,102 @@ def download_and_unpack_package(
+            elif protocol == Protocol.GS or protocol == Protocol.HTTPS:
+                # Download package via GS or HTTPS.
+                try:
+                    from smart_open import open
+                except ImportError:
+                    raise ImportError(
+                        "You must `pip install smart_open` "
+                        f"to fetch {protocol.value.upper()} URIs.")
+                with open(pkg_uri, "rb") as package_zip:
+                    with open(pkg_file, "wb") as fin:
+                        fin.write(package_zip.read())
+                unzip_package(pkg_file, local_dir, True, True, logger)
             else:
                 raise NotImplementedError(
                     f"Protocol {protocol} is not supported")
 
-            os.mkdir(local_dir)
-            logger.debug(f"Unpacking {pkg_file} to {local_dir}")
-            with ZipFile(str(pkg_file), "r") as zip_ref:
-                zip_ref.extractall(local_dir)
-            pkg_file.unlink()
-
         return str(local_dir)
 
 
+def get_top_level_dir_from_compressed_package(package_path: str):
+    """
+    If compressed package at package_path contains a single top-level
+    directory, returns the name of the top-level directory. Otherwise,
+    returns None.
+    """
+
+    package_zip = ZipFile(package_path, "r")
+    top_level_directory = None
+
+    for file_name in package_zip.namelist():
+        if top_level_directory is None:
+            # Set the top_level_directory when checking
+            # the first file in the zipped package
+            if "/" in file_name:
+                top_level_directory = file_name.split("/")[0]
+            else:
+                return None
+        else:
+            # Confirm that all other files
+            # belong to the same top_level_directory
+            if "/" not in file_name or \
+                    file_name.split("/")[0] != top_level_directory:
+                return None
+
+    # Ensure that zip file is not empty
+    return top_level_directory
+
+
+def unzip_package(package_path: str,
+                  target_dir: str,
+                  remove_top_level_directory: bool,
+                  unlink_zip: bool,
+                  logger: Optional[logging.Logger] = default_logger):
+    """
+    Unzip the compressed package contained at package_path and store the
+    contents in target_dir. If remove_top_level_directory is True, the function
+    will automatically remove the top_level_directory and store the contents
+    directly in target_dir. If unlink_zip is True, the function will unlink the
+    zip file stored at package_path.
+    """
+    try:
+        os.mkdir(target_dir)
+    except FileExistsError:
+        logger.info(f"Directory at {target_dir} already exists")
+
+    logger.debug(f"Unpacking {package_path} to {target_dir}")
+
+    if remove_top_level_directory:
+        top_level_directory = get_top_level_dir_from_compressed_package(
```

Comment:
as we discussed in slack, we can test this independently by crafting a zip file on local disk, assuming it's what we got from remote source and ensure everything works as we expect in an environment where we have complete control